### PR TITLE
TPM1.2: Modified NVReadValue so platform supplier need no auth in read.

### PR DESCRIPTION
**Pull request by ChriMarMe**

### CODE REVIEW

**Comment by ChriMarMe** [comment]
@tmroeder and here

**Comment by ChriMarMe** [comment]
Really need some review here please.

File: tpm/commands.go [hunk 1]
```diff
@@ -158,15 +158,24 @@ func getCapability(rw io.ReadWriter, cap, subcap uint32) ([]byte, error) {
 	return b, nil
 }
 
+//nvReadValue reads from the NVRAM
```
Comment (awly) [inline tpm/commands.go]: Add spaces between "//" and text

File: tpm/commands.go [hunk 2]
```diff
@@ -158,15 +158,24 @@ func getCapability(rw io.ReadWriter, cap, subcap uint32) ([]byte, error) {
...
+//If TPM isn't locked, no authentification is needed.
+//See TPM-Main-Part-3-Commands-20.4
 func nvReadValue(rw io.ReadWriter, index, offset, len uint32, ca *commandAuth) ([]byte, *responseAuth, uint32, error) {
 	var b tpmutil.U32Bytes
 	var ra responseAuth
-	in := []interface{}{index, offset, len, ca}
+	var in []interface{}
+	if ca != nil {
```
Comment (awly) [inline tpm/commands.go]:
```
in := []interface{}{index, offset, len}
if ca != nil {
  in = append(in, ca)
}
```

File: tpm/tpm.go [hunk 3]
```diff
@@ -929,6 +929,18 @@ func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth digest) ([
 	return data, nil
 }
 
+//NVReadValuePS returns the value from a given index, offset, and length in NVRAM.
```
Comment (awly) [inline tpm/tpm.go]: Spaces between // and text

File: tpm/tpm.go [hunk 4]
```diff
@@ -929,6 +929,18 @@ func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth digest) ([
...
+//If TPM isn't locked, no authentification is needed.
+//This is for platform suppliers only
+//See TPM-Main-Part-3-Commands-20.4
+func NVReadValuePS(rw io.ReadWriter, index, offset, len uint32) ([]byte, error) {
```
Comment (awly) [inline tpm/tpm.go]: Can you modify NVReadValue to support nil ownAuth instead?

File: tpm/tpm.go [hunk 5]
```diff
@@ -906,26 +906,41 @@ func ReadEKCert(rw io.ReadWriter, ownAuth digest) ([]byte, error) {
...
-		return nil, fmt.Errorf("failed to verify authenticity of response: %v", err)
+// If TPM isn't locked, no authentification is needed.
+// This is for platform suppliers only.
+// See TPM-Main-Part-3-Commands-20.4
+func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth []byte) ([]byte, error) {
+	var data []byte
+	var err error
+	var ra *responseAuth
+	var ret uint32
+	if ownAuth != nil {
```
Comment (awly) [inline tpm/tpm.go]:
```
if ownAuth == nil {
  data, _, _, err := nvReadValue(rw, index, offset, len, nil)
  if err != nil {
    return nil, fmt.Errorf("failed to read from NVRAM: %v", err)
  }
  return data, nil
}
// continue to non-nil ownAuth case without indentation
```

File: tpm/tpm.go [hunk 6]
```diff
@@ -906,26 +906,41 @@ func ReadEKCert(rw io.ReadWriter, ownAuth digest) ([]byte, error) {
...
+func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth []byte) ([]byte, error) {
+	var data []byte
+	var err error
+	var ra *responseAuth
+	var ret uint32
+	if ownAuth == nil {
+		data, _, _, err = nvReadValue(rw, index, offset, len, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read from NVRAM: %v", err)
+		}
```
Comment (awly) [inline tpm/tpm.go]: `return data, nil` after this and remove the else block below

File: tpm/tpm.go [hunk 7]
```diff
@@ -906,26 +906,41 @@ func ReadEKCert(rw io.ReadWriter, ownAuth digest) ([]byte, error) {
...
-		return nil, fmt.Errorf("failed to read from NVRAM: %v", err)
-	}
-	raIn := []interface{}{ret, ordNVReadValue, tpmutil.U32Bytes(data)}
-	if err := ra.verify(ca.NonceOdd, sharedSecretOwn[:], raIn); err != nil {
-		return nil, fmt.Errorf("failed to verify authenticity of response: %v", err)
+// If TPM isn't locked, no authentification is needed.
+// This is for platform suppliers only.
+// See TPM-Main-Part-3-Commands-20.4
+func NVReadValue(rw io.ReadWriter, index, offset, len uint32, ownAuth []byte) ([]byte, error) {
+	var data []byte
```
Comment (awly) [inline tpm/tpm.go]: Remove all these var declarations, use `:=` as appropriate below

**Pull request merged**